### PR TITLE
Add CVE-2024-38077

### PR DESCRIPTION
**Pull request by nmasdoufi-ol**

### CODE REVIEW

File: tests/exploits/cve_2024_38077_test.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+"""Unit tests for Agent Asteroid: CVE-2024-38077"""
+
+from unittest import mock
```
Comment (benyissa) [inline tests/exploits/cve_2024_38077_test.py:3]: I would use pytest_mock as we did in previous tests for consistency.

File: agent/exploits/cve_2024_38077.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
+"""Agent Asteroid implementation for CVE-2024-38077"""
+
+import logging
+from impacket.dcerpc.v5 import transport, epm  # type: ignore
+from impacket.dcerpc.v5.ndr import NDRSTRUCT, NDRCALL  # type: ignore
+from impacket.dcerpc.v5.dtypes import PULONG  # type: ignore
+from impacket.dcerpc.v5.rpcrt import DCERPCException  # type: ignore
+from impacket.uuid import uuidtup_to_bin  # type: ignore
```
Comment (elyousfi5) [inline agent/exploits/cve_2024_38077.py:8]:
```suggestion
from impacket import uuid  # type: ignore
```

File: agent/exploits/cve_2024_38077.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+from ostorlab.agent.kb import kb
+from ostorlab.agent.mixins import agent_report_vulnerability_mixin
+
+from agent import definitions
+from agent import exploits_registry
+
+UUID = uuidtup_to_bin(("3d267954-eeb7-11d1-b94e-00c04fa3080d", "1.0"))
```
Comment (elyousfi5) [inline agent/exploits/cve_2024_38077.py:16]:
```suggestion
UUID = uuid.uuidtup_to_bin(("3d267954-eeb7-11d1-b94e-00c04fa3080d", "1.0"))
```

File: agent/exploits/cve_2024_38077.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def _connect_to_license_server(target_ip: str) -> str | None:
+    try:
+        stringbinding = epm.hept_map(target_ip, UUID, protocol="ncacn_ip_tcp")
+        logger.info("RPC Binding String: %s", stringbinding)
+        rpctransport = transport.DCERPCTransportFactory(stringbinding)
+        rpctransport.set_connect_timeout(30)
+        dce = rpctransport.get_dce_rpc()
+        dce.set_auth_level(2)
```
Comment (elyousfi5) [inline agent/exploits/cve_2024_38077.py:58]: please put numbers as consts

File: agent/exploits/cve_2024_38077.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+
+class ContextHandle(NDRSTRUCT):  # type: ignore
+    """Represents the context handle structure used in RPC calls."""
+
+    structure = (("Data", "20s=b"),)
+
+    def getAlignment(self) -> int:
+        """Returns the alignment value for the structure."""
+        return 4
```
Comment (mohsinenar) [inline agent/exploits/cve_2024_38077.py:37]: Put in a const!

**Review by mohsinenar — CHANGES_REQUESTED** [review]

File: agent/exploits/cve_2024_38077.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+import logging
+from impacket.dcerpc.v5 import transport, epm  # type: ignore
+from impacket.dcerpc.v5.ndr import NDRSTRUCT, NDRCALL  # type: ignore
+from impacket.dcerpc.v5.dtypes import PULONG  # type: ignore
+from impacket.dcerpc.v5.rpcrt import DCERPCException  # type: ignore
+from impacket import uuid  # type: ignore
+
+from ostorlab.agent.kb import kb
+from ostorlab.agent.mixins import agent_report_vulnerability_mixin
```
Comment (rabsondev) [inline agent/exploits/cve_2024_38077.py:11]:
```suggestion
import logging

from impacket.dcerpc.v5 import transport, epm  # type: ignore
from impacket.dcerpc.v5.ndr import NDRSTRUCT, NDRCALL  # type: ignore
from impacket.dcerpc.v5.dtypes import PULONG  # type: ignore
from impacket.dcerpc.v5.rpcrt import DCERPCException  # type: ignore
from impacket import uuid  # type: ignore
from ostorlab.agent.kb import kb
from ostorlab.agent.mixins import agent_report_vulnerability_mixin
```

File: agent/exploits/cve_2024_38077.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        get_version["version"] = 3
+        res_get_version = dce.request(get_version)
+        version = res_get_version["version"]
+        return "0x{:x}".format(version)
+
+    except DCERPCException as e:
+        logger.error("Failed to connect or get version: %s", e)
+        return None
+
+    except Exception as e:
```
Comment (rabsondev) [inline agent/exploits/cve_2024_38077.py:96]: Is this intentional (catching the generic exception)?

File: agent/exploits/cve_2024_38077.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        """Checks if the target is likely vulnerable by attempting to connect.
+
+        Args:
+            target (definitions.Target): The target to check.
+
+        Returns:
+            bool: True if the target is likely vulnerable, False otherwise.
+        """
+        try:
+            return _connect_to_license_server(target.origin) is not None
```
Comment (rabsondev) [inline agent/exploits/cve_2024_38077.py:155]: The function catches `Exception` so why is there a try/except?

**Review by rabsondev — CHANGES_REQUESTED** [review]

File: agent/exploits/cve_2024_38077.py (hunk 8 above)
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_38077.py:156]: Some unit tests were failing, that's why I had to add it.

Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_38077.py]: I fixed it.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: agent/exploits/cve_2024_38077.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
+"""Agent Asteroid implementation for CVE-2024-38077"""
+
+import logging
+
+from impacket.dcerpc.v5 import transport, epm  # type: ignore
+from impacket.dcerpc.v5.ndr import NDRSTRUCT, NDRCALL  # type: ignore
+from impacket.dcerpc.v5.dtypes import PULONG  # type: ignore
+from impacket.dcerpc.v5.rpcrt import DCERPCException  # type: ignore
```
[the remark this answers is not in the document]
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_38077.py]: Done.

File: agent/exploits/cve_2024_38077.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    def accept(self, target: definitions.Target) -> bool:
+        """Checks if the target is likely vulnerable by attempting to connect.
+
+        Args:
+            target (definitions.Target): The target to check.
+
+        Returns:
+            bool: True if the target is likely vulnerable, False otherwise.
+        """
+        return _connect_to_license_server(target.origin) is not None
```
[the remark this answers is not in the document]
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_38077.py:149]: Thanks  🙏

File: agent/exploits/cve_2024_38077.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
+"""Agent Asteroid implementation for CVE-2024-38077"""
+
+import logging
+import socket
+
+from impacket.dcerpc.v5 import transport, epm  # type: ignore
+from impacket.dcerpc.v5 import ndr, dtypes, rpcrt
```
Comment (3asm) [inline agent/exploits/cve_2024_38077.py:7]: Split import, not on the same line.

**Review by 3asm — APPROVED** [review]

**Pull request merged**